### ISSUE TITLE
Consider allowing subtracting / adding quantities with the same physical dimension but different unit prefixes

### ISSUE THREAD

**Issue by jokasimr** [scipp/scipp#3456]
## TLDR:
* Adding variables of the same physical dimension but with different unit prefixes is not allowed.
* The reason is concern about precision loss, but unit prefixes being different is not sufficient or necessary for precision loss.
* In my opinion it's a bad heuristic with high false positive rate.
* Scipp user experience would be improved by not raising unit error in this case.

## Issue description
Currently none of these are allowed:
```python
sc.scalar(1.0, unit='deg') + sc.scalar(1.0, unit='rad')
sc.scalar(1.0, unit='m') + sc.scalar(1.0, unit='mm')
```
The reason for not allowing this is that we are concerned about precision loss.
For example, in a case like this
```python
sc.scalar(1.3, unit='m', dtype='float32') + sc.scalar(2.4, unit='angstrom', dtype='float32')
```
the result will just be rounded to `1.3m`.
However, provided that the user really intended to add the above quantities, the result is still correct in the sense that it is correct to the precision of the data type.
A worse example would be:
```python
sc.scalar(1.3, unit='m', dtype='float32') + sc.scalar(2.4, unit='angstrom', dtype='float32') - sc.scalar(1.3, unit='m', dtype='float32')
```
Here the result is just wrong, the user needs to reorder the operations. If we raise the `UnitError` here that might catch their attention that something is fishy and needs fixing. However, if they just fix it by doing:
```python
sc.scalar(1.3, unit='m', dtype='float32') + sc.scalar(2.4, unit='angstrom', dtype='float32').to(unit='m') - sc.scalar(1.3, unit='m', dtype='float32')
```
the result is still wrong! In other words it's not the case that the unit system accidentally *causes* precision loss. Units being different is used as a *heuristic* that precision might be lost.

Is the `UnitError` a good heuristic for this purpose? I don't think it is, because in the majority of cases when it is triggered there is no risk of precision loss. It has a high false positive rate. And since we can still get precision problems just because the units are the same, it has a significant false negative rate too.

The unit conversions makes code more verbose. Example from the reflectometry workflow:
```python
def angular_resolution(
    da: FootprintCorrectedData[Sample],
    detector_spatial_resolution: DetectorSpatialResolution[Sample],
) -> AngularResolution:
    theta = da.bins.coords['theta']
    return (
        fwhm_to_std(
            sc.to_unit(
                sc.atan(
                    sc.to_unit(detector_spatial_resolution, 'm')
                    / sc.to_unit(da.coords['position'].fields.z, 'm', copy=False)
                ),
                theta.bins.unit,
                copy=False,
            )
        )
        / theta
    )
```
without unit conversions this would read like:
```python
def angular_resolution(
    da: FootprintCorrectedData[Sample],
    detector_spatial_resolution: DetectorSpatialResolution[Sample],
) -> AngularResolution:
    theta = da.bins.coords['theta']
    return (
        fwhm_to_std(
            sc.atan(detector_spatial_resolution /
           da.coords['position'].fields.z,
        )
        / theta
    )
```
(slightly unfair comparison because the first version also avoids copying in the unit conversion, but I think the point is clear anyway).

**Comment by SimonHeybrock** [comment]
> The reason for not allowing this is that we are concerned about precision loss.

Note that there are more reasons:
- Hidden behavior that may be hard to spot ("why is my data suddenly in mm") in a longer workflow.
- Hidden performance penalty, e.g., from converting the "wrong" operand, requiring large copy/memory use. While avoidable in principle by doing it during a binary operation, this is not possible in the current implementation, i.e., we need to call the conversion *separately* from the binary operation.

Aspects to consider:
- What should the result unit be? First operand? Larger unit? Smaller unit? Unit closer to SI?
- Why limit to `+` and `-`? What about all the other operations?

**Comment by jokasimr** [comment]
> What should the result unit be?

We could do the same as is done by Pint, that is, First operand. Or, operand with smallest size, to avoid the 'Hidden performance penalty' issue.

> Why limit to + and -? What about all the other operations?

`*` and `/` doesn't have the same problem. But yes let's do it for all operands where it makes sense.

> Hidden performance penalty, e.g., from converting the "wrong" operand, requiring large copy/memory use. 

Hm, yes it is a problem. The user will have to convert explicitly to avoid it, but that is sometimes hard to do in practice (if you don't know ahead of time what operand is larger).

**Comment by SimonHeybrock** [comment]
> > What should the result unit be?
> 
> We could do the same as is done by Pint, that is, First operand. Or, operand with smallest size, to avoid the 'Hidden performance penalty' issue.
> 
> > Why limit to + and -? What about all the other operations?
> 
> `*` and `/` doesn't have the same problem. But yes let's do it for all operands where it makes sense.

It does to some extent, e.g., dividing `m/km` returns a dimensionless unit with a scale factor, which is really surprising and also has UX problems.

**Comment by nvaytet** [comment]
I would be in favor of not allowing this. I don't think it's such a pain to have to convert manually, and I feel more in control of my operations, as opposed to let the library do its best guess.
Sometimes, I end up with strange combinations with pint, especially when you involve units like e.g. angstroms.

> dividing m/km returns a dimensionless unit with a scale factor, which is really surprising and also has UX problems.

I also don't like this. If we could fix that somehow, it would be great, but I don't know what 'fixing' here would mean.

**Comment by jokasimr** [comment]
> I don't think it's such a pain to have to convert manually, and I feel more in control of my operations, as opposed to let the library do its best guess.

Converting manually will always be an option anyway. The question is if we make it optional or mandatory.

**Comment by SimonHeybrock** [comment]
Note that `sc.to_unit` is more verbose than using the `to` method:
```python
def angular_resolution(
    da: FootprintCorrectedData[Sample],
    detector_spatial_resolution: DetectorSpatialResolution[Sample],
) -> AngularResolution:
    theta = da.bins.coords['theta']
    return (
        fwhm_to_std(
            sc.atan(
                detector_spatial_resolution.to(unit='m')
                da.coords['position'].fields.z.to(unit='m', copy=False)
            ).to(unit=theta.bins.unit, copy=False)
        )
        / theta
    )
```

**Comment by jokasimr** [comment]
I took a look at how other libraries with physical units handles this. The conclusion is that every library I looked at does allow adding units of the same dimension but with different prefixes. That indicates others have concluded it to be a sensible design.

Libraries investigated:
* [Pint](https://pint.readthedocs.io/en/stable/getting/tutorial.html) (python)
* [Unyt](https://unyt.readthedocs.io/en/stable/) (python)
* [Astropy.unit](https://het.as.utexas.edu/HET/Software/Astropy-1.0/units/index.html) (python)
* [Unitful.jl](https://painterqubits.github.io/Unitful.jl/stable/conversion/#ContextUnits) (julia)
* [mp-unit](https://github.com/mpusz/mp-units?tab=readme-ov-file#tldr) (c++)

I think it's important to consider that this is not only a convenience issue. It's also about composability.
When code needs defensive unit conversions it becomes more difficult to reuse code that wasn't written with units in mind, and vice versa.

**Comment by SimonHeybrock** [comment]
> > dividing m/km returns a dimensionless unit with a scale factor, which is really surprising and also has UX problems.
> 
> I also don't like this. If we could fix that somehow, it would be great, but I don't know what 'fixing' here would mean.

I think we would probably want to move any pure scale factor from the unit to the values?

**Comment by damskii9992** [comment]
Since we are starting to use Scipp to hold values and units internally in the EasyScience framework, I encountered this surprising behavior as well. And I personally think @jokasimr's comment about how all other unit handling libraries are doing it is a very valid reason to allow this. Since it is what most users (including myself) would expect, coming from other frameworks and libraries.

A solution, albeit likely an undesirable one, to the precision loss, would be to do what I was taught in my error propagation classes in university, which is to treat any number, 2.0 as 2.00 \u00B 0.05 where you can then add as many trailing 0's as you want. The arithmetic operation:
```python
sc.scalar(1.3, unit='m', dtype='float32') + sc.scalar(2.4, unit='angstrom', dtype='float32') - sc.scalar(1.3, unit='m', dtype='float32')
```
would then result in a value of 2.4 Å with a very large uncertainty of 0.1 m (or 1 000 000 Å). Similarly holds for additions such as 1.0 + 0.01, which should then also add such an uncertainty.

Regardless of the above proposal, I would still allow the addition/subtraction of same dimension values, simply because it is what users will expect coming from other softwares/frameworks. In EasyScience, we will (for now) be taking the Pint approach of converting the 2nd operand unit to allow this, until this is changed in Scipp.
